### PR TITLE
Add Transitive Dependencies Spec

### PR DESCRIPTION
**Pull request by JonDouglas**

### CODE REVIEW

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md:15]: "...promoting **_to_** a top-level package will resolve a conflict"?

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md:73]: > what top-level package(s) are bringing it in.

If there are no top-level packages bringing it in, would the comment simply reflect this? e.g. "This transitive dependency is brought in by one or more transitive dependencies."

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md:124]: I like the sound of "promote", as you're right--I think "installed" misleads people into thinking it's not installed already.

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md:123]: My preference would be "direct dependencies" and "indirect dependencies" here, if we're going for more simple language. I think this is descriptive enough without being too esoteric.

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md:139]: This exists in CLI already.

**Review by drewgillies — APPROVED** [review]
A few minor tweaks/thoughts, but looks good, @JonDouglas! Thanks. :shipit:

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [inline designs/Transitive-Dependencies.md]: Note--this is a NuGet link. Is that intentional?

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [inline designs/Transitive-Dependencies.md:55]: Sounds like we are also trying to solve https://github.com/NuGet/Home/issues/6339 with this. 

Visual Studio 2015 had a Package Visualizer you can look into.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [inline designs/Transitive-Dependencies.md:70]: Are we going to audit transitive packages automatically and display that in the updates/installed labels? 

That's what we're doing for the top level packages. 

Note that I'm not recommending this, merely just clarifying :)

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [inline designs/Transitive-Dependencies.md:80]: I found it weird that the goals are below the what. :) 

Maybe there was a meeting benefit here, but I think I was OOF that day.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:123]: I've seen the argument both directions, particularly top level vs transitive and direct vs indirect and in the end I think consistency matters.

My suggestion would be to double down to what we've already committed in top level and transitive. 

Note that other package managers such as paket, pub, npm already use top-level/direct & transitive. 

I've rarely seen indirect being used.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:124]: I agree with the promote suggestion here.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [inline designs/Transitive-Dependencies.md:150]: When installing, licenses from *all* packages being installed, including transitive dependencies are displayed in the license dialog.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [inline designs/Transitive-Dependencies.md:74]: It's not clear to me how the dependency path for a given transitive dependency would work based on this screenshot. 

I just 2 separate lists.

**Review by nkolev92 — COMMENTED** [review]
I was OOF when the original meeting happened, so I might have some more basic questions :)

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:74]: @nkolev92 There's a tooltip control in the screenshot that will show the dependency relation of a given transitive dependency. This however is likely to be cut given it's not very accessible :/

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:150]: Thanks for the clarification! Nothing changes in this regard then.

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:150]: For "Promote" we are awaiting another UX workshop to go over these small details such as what language to be used. We want to remain consistent, but promote probably has a better connotation than "installing" it. Thanks to @skofman1 for the suggestion!

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:150]: Top-level and transitive are things I don't believe we can change although the UX team would like to review them as they are "stop-and-think" words.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:80]: Haha, I just wanted to make sure people were reminded of the goals, but more importantly just to blabber non-goals 👍

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:70]: Mind clarifying a bit more on "audit transitive packages automatically"? Just so I understand further!

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:55]: I knew there was something... I think this is a separate thing entirely and can remain in that issue 💯

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:123]: UX Board will review this, but I have been a broken record of using language everyone already knows and is documented.

https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files for example uses top-level & transitive but it also uses direct. 🕺

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:123]: https://docs.microsoft.com/en-us/nuget/what-is-nuget also uses top-level & down-level

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:124]: 👍

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:73]: I'm not quite sure. How does that work exactly? Do we have an example where a transitive-only situation happens? It has to be brought in by at least one top-level package right?

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md]: Copy / pasted a template and never updated it.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:123]: `down-level`
🤕

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:74]: Ah, I missed that.

File: designs/Transitive-Dependencies.md
Comment (drewgillies) [reply designs/Transitive-Dependencies.md:73]: Good point. Ultimately it comes from a top-level package. I was thinking of the scenario like showing package D's details here:
[only toplevel package]A->B
B->C
C->M
M->N
M->D
N->D
From what you're saying, you'd list A as the top-level package bringing in D, and that's it. I was reading your text and thinking: since D _isn't_ brought in directly by a top-level package (directly being the operative word), what would you list? So you'd go with A here. Makes sense. The way a customer would use this information is--if I get rid of A, I get rid of D. 👍

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:73]: Yep! Just showing the root of the package that ultimately brought the transitive dependency in. (could be many!)

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:125]: Can the reasons why be elaborated? It may have been mentioned in a meeting, but I can't recall and think the details should be here.

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:66]: Accessibility should be called out. If we have any input from the a11y team, let's add it as well. At a minimum, note here that we will run checks in Accessibility Insights, and ensure screen-readers can differentiate Top/Transitive, as well as read the tooltip content.

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:68]: Is this spec committing to this specific implementation within PM UI? If not, the wording should be changed as such. Also consider whether there will be an iterative approach to bringing transitives to the UI, and what those phases might look like.

I think the collapsing panel is a fine paradigm, but did we consider alternatives for tops vs transitives? 
Perhaps a filter option to "Show transitives" would work just as well? 
The description in https://github.com/NuGet/Home/issues/5887 shows a screenshot with a highlight around a package that's "installed explicitly"/top. This made me think that a visual indicator (color/etc) and screen-reader property may work just as well. And much like our `show prerelease`, they can be filtered out.

What do we gain from a collapsible panel? If there's always going to be on the order of N=10 top-levels, and N=100's of Transitives, it's hard to envision why I would want the collapsing behavior.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:70]: Currently with the top level only approach, the updates tab will have a label telling you the amount of packages that have updates (this is pretty bad perf right now). 

When deprecation was added, a warning icon was equivalent to the number in the Updates tab on the Installed tab.
This means that when we look-up the metadata for the packages installed to determine if they have any newer versions, we also look at the metadata to determine whether they are deprecated. 

Are we going to do the equivalent for transitive dependencies?
Are we going to preemptively look for the metadata for the transitive dependencies to determine if they are deprecated/vulnerable? 

In particular I'm asking this because given the way the server protocol is written right now, this could lead to 100s of different network calls.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:68]: Yes this spec is committing to this implementation as there was really no suggested alternative & people generally agreed this was a good approach. If we don't feel that way today, we should reiterate.

This work by itself is phased from the package vulnerability work. Because of the scope of this work, there's really no phased approach to this specific design.

The reason there is a collapsing concept is to allow users to have their screen real estate back if they do not care for transitive dependencies and would like to keep them collapsed while doing top-level management tasks.

The decision for two separate lists was primarily driven by many different constructs such as the existing dotnet CLI behavior for showing transitives, and the many eco systems that show top-level dependencies & transitive dependencies in separate lists. (gradle, ant, etc). Others show dependency trees, which was not designed previously and thus we're committing to a similar experience already completed in dotnet.

The jobs to be done are to look at top-level, transitive level, or both. There are many reasons one may look at each level & having collapsable controls helps users decide how they'd like their screen to be shown. Otherwise yes, we could do a filter, but the purpose was to differentiate the types of references clearly and not have them all in one list with a UI affordance to say "transitive" or "top-level" although you might be able to filter/order by that.

**Comment by drewgillies** [comment]
@JonDouglas @donnie-msft @nkolev92 Conversation seems to have ceased on this spec. Shall we merge it and call it adopted?

**Comment by nkolev92** [comment]
I think there was some feedback from me & @donnie-msft that might need addressing. 
I was just waiting for a ping to re-read when that's done (re-review notification basically).

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:15]: Resolving via new commits.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:123]: We are keeping the language top-level / transitive to be consistent. If UX board believes we should change it at a craftsman review, then we can reconsider.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md]: Updating via commit

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:70]: What is the perf potential of 100s of network calls?

You would see the same behavior for any installed package whether top-level or transitive.

Installed tab would show the current version & the available version.

Update tab would show the current version & the available version.

We would only count top-level packages for the badge on the tabs given we don't want users to promote to top-level just for a new version.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:125]: 1. .NET Core has accessibility properties for Tooltip, .NET Framework supposedly doesn't.
2. Tooltip doesn't scale well with listing potentially hundreds of packages that may bring in the transitive dependency.
3. UX Board wanted to introduce a new concept, but hasn't scheduled a workshop yet.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:66]: Accessibility concerns should've been called out regarding the tooltip aspect. We will do a full compatibility pass when we're nearing the craftsmanship review of the UX board. It's likely we will have to cut the tooltip given we have no alternatives & need to unblock package vulnerability work for now.

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [reply designs/Transitive-Dependencies.md:125]: Was resolved but I don't see the details added to the spec

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [reply designs/Transitive-Dependencies.md:66]: Was resolved but I don't see the details added to the spec. If there will be a technical spec coming after this, let me know.

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:66]: I didn't add any details to this spec as it's a design document. Although accessibility is considered & called out in this spec already, we don't have any insight or clarity into further accessibility concerns right now & those may need to be considered in the technical implementation as these are more test/implementation scenarios than design scenarios.

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:66]: /cc @drewgillies if you were planning to write-up anything for technical spec to make note of accessibility during implementation.

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:95]: The spec feels a bit more PM level, but now we're getting specific with XAML controls. Are these just high level concepts of what we think we're changing in the UI?
When you say ScrollViewer... you expect a new scrollbar so that Top level and Transitive level have their own individual scrollbars?

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:99]: Mentioning the context from the other thread regarding a11y concerns, etc, would be good here.
Actually since you say below that design team has not yet decided this, we should indicate it as Optional for now?

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [inline designs/Transitive-Dependencies.md:100]:
```suggestion
The above elements will be checked against Accessibility Insights. We'll ensure screen-readers can differentiate Top versus Transitive dependencies, and the scrolling and expansion functionality will be accessible with the keyboard.
```

**Review by donnie-msft — CHANGES_REQUESTED** [review]

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [reply designs/Transitive-Dependencies.md:125]: Moved this concern to the Elements section.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:95]: I have no concept of what controls should be used. I defer to the individual/implementation to use the best controls. These are just three elements I'm aware of to accomplish the design today.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:99]: Thanks for the commit. This should be deferred to the implementation & the a11y concerns for whatever controls are ultimately used.

File: designs/Transitive-Dependencies.md
Comment (donnie-msft) [reply designs/Transitive-Dependencies.md:95]: Ok, then I would think we shouldn't list them in the spec if they are being deferred.

**Review by donnie-msft — APPROVED** [review]
Looks good. Taking the specific implementation notes with a grain of salt. Suggested removing them as they seem to be authoritative.

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:70]: We usually throttle based on the CPUs. 
The impact would be delayed population of the information in the UI. 

The number of network calls NuGet makes to load the information needed for UI is a problem today for the solution tab where the package count is large. A rough guestimate would be that the project level PM UI now needs as much information as the solution level one. 

If we don't want to shown badges for top level, then these concerns go away :) 

Note that these perf problems should be somewhat solvable, it's just a resourcing concern that'd likely take more effort than just displaying the transitive dependencies :)

**Review by nkolev92 — APPROVED** [review]
@donnie-msft brought up some good concerns, but beyond that + the thread I just replied on looks good!

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:70]: It's also completely fine to just not show updates available for transitive either. It would be a nice parity thing to reuse the current experience to show if a resolved transitive has an update with the hover experience. If it's complex perf wise, we can cut it.

File: designs/Transitive-Dependencies.md
Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:95]: Added a sentence to help distinguish!

File: designs/Transitive-Dependencies.md
Comment (nkolev92) [reply designs/Transitive-Dependencies.md:70]: I think that's perfectly reasonable. 

I basically want us to make progress on the basics in this area without having to worry about perf implications now. 

If we want to have full parity later, we'll judge it on it's own merit.

Comment (drewgillies) [reply designs/Transitive-Dependencies.md:70]: The whole point of implementing transitive dependencies is for the metadata, right? Don't we want to see which transitive dependencies are vulnerable?

Comment (drewgillies) [reply designs/Transitive-Dependencies.md:70]: My understanding at this time, is that we're chiefly implementing transitive dependencies in PMUI *specifically* for the metadata, right? Isn't the North Star here being able to see which transitive packages have vulnerabilities?

Comment (drewgillies) [reply designs/Transitive-Dependencies.md:70]: (I do know that's not a part of this spec, but since it's come up in conversation, and we want a design that supports these future features, let's discuss it)

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:70]: I'm deferring to you both regarding what the complexities would be. The goal is to show transitive dependencies enough to unblock package vulnerability while keeping a consistent experience that we have today  🏃

Comment (nkolev92) [reply designs/Transitive-Dependencies.md:70]: The `perf` concern I have is if we aggressively fetch the metadata. 

ie, do we fetch metadata for transitive dependencies if the transitive dependencies are collapsed?

Right now, if you go to the browse tab, we'll fetch the metadata for all packages on that view...just so we can show the label.

Comment (drewgillies) [reply designs/Transitive-Dependencies.md:70]: I'm looking to take a step back and consider the intention driving this feature, and whether we're still achieving this objective by not fetching metadata for a collapsed list. If a customer is viewing the installed packages for their project and sees a collapsed transitive dependencies list, and we haven't yet parsed vulnerability or deprecation metadata for the dependency graph, the customer may be misled into thinking the project's only vulnerabilities (for example) are in the top-level packages, which brings us back to the product's trustworthiness. 

The way I see it, there are 3 paths forward:
- Leave transitive dependency metadata unfetched, transitive dependencies collapsed, therefore incomplete vulnerability/deprecation data for the customer (until the list is expanded). We could add a disclaimer on the collapsed list indicating that expanding the list is the only way to get a complete picture of project deprecation/vulnerability status, and that this may take some time to fetch. This will however give an inaccurate label on the tab if we wish to indicate vulnerability/deprecation status there.

- Fetch all the data for the graph up-front and take the perf hit.

- Rearchitect our services approach to provide a more performant (complete) picture up front.

Comment (JonDouglas) [reply designs/Transitive-Dependencies.md:70]: I'm going to merge this spec as per other approvals & we'll take up these questions at implementation or technical spec/design time.

**Pull request merged**
